File: book/swidgets/c#/SWidgets/SComboBox.cs
using System;
using System.Collections.Generic;
using System.Windows.Controls;
using Sodium.Frp;
using Sodium.Functional;

namespace SWidgets
{
    public class SComboBox<T> : ComboBox, IDisposable
    {
        private readonly Action disposeListeners;

        public SComboBox()
            : this(null)
        {
        }

        public SComboBox(IEnumerable<T> items)
            : this(Stream.Never<Maybe<T>>(), Maybe.None, items)
        {
        }

        public SComboBox(Maybe<T> initSelectedItem)
            : this(initSelectedItem, null)
        {
        }

        public SComboBox(Maybe<T> initSelectedItem, IEnumerable<T> items)
            : this(Stream.Never<Maybe<T>>(), initSelectedItem, items)
        {
        }

        public SComboBox(Stream<Maybe<T>> setSelectedItem, Maybe<T> initSelectedItem)
            : this(setSelectedItem, initSelectedItem, null)
        {
        }

        public SComboBox(Stream<Maybe<T>> setSelectedItem, Maybe<T> initSelectedItem, IEnumerable<T> items)
        {
            void SetSelectedItemImpl(Maybe<T> m) => base.SelectedItem = m.Match<object>(v => v, () => null);

            this.ItemsSource = items ?? new T[0];
            SetSelectedItemImpl(initSelectedItem);

            List<IListener> listeners = new List<IListener>();

            StreamSink<int> sDecrement = Stream.CreateSink<int>();
            Cell<bool> allow = setSelectedItem.Map(_ => 1).OrElse(sDecrement).Accum(0, (b, d) => b + d).Map(b => b == 0);

            Maybe<T> GetSelectedItem()
            {
                object sel = base.SelectedItem;
                return sel == null ? Maybe.None : Maybe.Some((T)sel);
            }

            StreamSink<Maybe<T>> sUserSelectedItem = Stream.CreateSink<Maybe<T>>();
            this.SUserSelectedItem = sUserSelectedItem;
            this.SelectedItem = sUserSelectedItem.Gate(allow).OrElse(setSelectedItem).Hold(initSelectedItem);

            void SelectionChangedEventHandler(object sender, SelectionChangedEventArgs args)
            {
                Maybe<T> selectedItem = GetSelectedItem();
                this.Dispatcher.InvokeAsync(() => sUserSelectedItem.Send(selectedItem));
            }

            this.SelectionChanged += SelectionChangedEventHandler;

            listeners.Add(setSelectedItem.Listen(m =>
            {
                this.Dispatcher.InvokeAsync(() =>
                {
                    this.SelectionChanged -= SelectionChangedEventHandler;
                    SetSelectedItemImpl(m);
                    this.SelectionChanged += SelectionChangedEventHandler;
                    sDecrement.Send(-1);
                });
            }));

            this.disposeListeners = () =>
            {
                foreach (IListener l in listeners)
                {
                    l.Unlisten();
                }
            };
        }

        public new Cell<Maybe<T>> SelectedItem { get; }
        public Stream<Maybe<T>> SUserSelectedItem { get; }

        public void Dispose()
        {
            this.disposeListeners();
        }
    }
}

File: book/swidgets/f#/SWidgets/SComboBox.fs
namespace SWidgets

open System
open System.Windows.Controls
open Sodium.Frp

type SComboBox<'T>(setSelectedItem : Stream<'T option>, initSelectedItem : 'T option, items : 'T seq) as this =
    inherit ComboBox()

    let mutable selectionChangedEventHandler : IDisposable option = None

    let getObjectFromSelectedItem m =
        (match m with
            | None -> null
            | Some v -> box v)

    let init () =
        let sDecrement = sinkS ()
        let allow = (setSelectedItem |> mapToS 1, sDecrement) |> orElseS |> accumS 0 (+) |> mapC ((=) 0)
        let getSelectedItem () =
            let sel = this.BaseSelectedItem
            if sel = null then None else match (box sel) with | :? 'T as s -> Some s | _ -> None
        let sUserSelectedItem = sinkS ()
        let selectedItem = (sUserSelectedItem |> Stream.gate allow, setSelectedItem) |> orElseS |> holdS initSelectedItem
        let subscribe () = 
            this.SelectionChanged.Subscribe (fun _ ->
                let selectedItem = getSelectedItem ()
                this.Dispatcher.InvokeAsync(fun () -> sUserSelectedItem |> sendS selectedItem) |> ignore)
        let listener = setSelectedItem |> Stream.listen (fun o ->
            this.Dispatcher.InvokeAsync(fun () ->
                match selectionChangedEventHandler with | None -> () | Some h -> h.Dispose()
                this.BaseSelectedItem <- getObjectFromSelectedItem o
                selectionChangedEventHandler <- Option.Some (subscribe ())
                sDecrement |> sendS -1) |> ignore)
        sUserSelectedItem, selectedItem, subscribe, (fun () -> listener.Unlisten ())

    let sUserSelectedItem, selectedItem, subscribe, disposeListener = init ()

    do
        base.ItemsSource <- items
        base.SelectedItem <- getObjectFromSelectedItem initSelectedItem
        selectionChangedEventHandler <- Option.Some (subscribe ())

    new(setSelectedItem, initSelectedItem) = new SComboBox<_>(setSelectedItem, initSelectedItem, Seq.empty)
    new(initSelectedItem, items) = new SComboBox<_>(Stream.never (), initSelectedItem, items)
    new(initSelectedItem) = new SComboBox<_>(Stream.never (), initSelectedItem)
    new(items) = new SComboBox<_>(Stream.never (), Option.None, items)
    new() = new SComboBox<_>(Seq.empty)

    member private __.BaseSelectedItem
        with get () = base.SelectedItem
        and set value = base.SelectedItem <- value

    member val SelectedItem = selectedItem
    member val SUserSelectedItem = sUserSelectedItem

    interface IDisposable with
        member __.Dispose() = disposeListener ()